[mlir][emitc] Mark emitc.literal with CExpression

Vladislave0-0 opened the pull request:
This PR marks `emitc.literal` as a valid operation within `emitc.expression`, removing an artificial restriction since literals are inherently inlined into expressions during emission.

Example:
```mlir
%0 = emitc.expression : i32 {
  %literal = emitc.literal "42" : i32
  emitc.yield %literal : i32
}
```

Vladislave0-0 committed:
[mlir][emitc] Mark emitc.literal with CExpression

This PR marks `emitc.literal` as a valid operation within
`emitc.expression`, removing an artificial restriction since
literals are inherently inlined into expressions during emission.

Example:
```mlir
%0 = emitc.expression : i32 {
  %literal = emitc.literal "42" : i32
  emitc.yield %literal : i32
}
```

marbre commented on mlir/lib/Target/Cpp/TranslateToCpp.cpp:
```diff
@@ -98,6 +98,7 @@ static FailureOr<int> getOperatorPrecedence(Operation *operation) {
       .Case<emitc::ConditionalOp>([&](auto op) { return 2; })
       .Case<emitc::DivOp>([&](auto op) { return 13; })
       .Case<emitc::LoadOp>([&](auto op) { return 16; })
+      .Case<emitc::LiteralOp>([&](auto op) { return 16; })
```
LGTM technically but I am not sure if this is the correct precedence for the LiteralOp (see https://en.cppreference.com/w/cpp/language/operator_precedence.html and https://en.cppreference.com/w/c/language/operator_precedence.html).

Vladislave0-0 committed:
[mlir][emitc] Changes after review

Vladislave0-0 commented on mlir/test/Target/Cpp/expressions.mlir:
```diff
@@ -439,3 +439,29 @@ emitc.func @expression_with_call_opaque_with_args_array(%0 : i32, %1 : i32) {
+// CPP-DECLTOP-NEXT: }
+
+func.func @expression_with_literal(%arg0 : i32) -> i1 {
+  %ret = emitc.expression noinline : i1 {
```
> IIRC the literals itself doesn't care about precedence, so it acts Like having highest precedence and this (or maybe 17?) is fine.

Thank you. I also followed the logic that `LiteralOp` should have a maximum precedence. I think 17 will really be better.

> Literals are Always inlined. Can we have a Test that wraps a single literal and that gets inlined into it's use?
Like return(expression(literal(...)))

I sent this test in a new commit. In this case, does the `LiteralOp` really need the  `Pure` trait?

Vladislave0-0 commented on mlir/include/mlir/Dialect/EmitC/IR/EmitC.td:
```diff
@@ -851,7 +851,7 @@ def EmitC_IncludeOp
 }
 
-def EmitC_LiteralOp : EmitC_Op<"literal", [Pure]> {
+def EmitC_LiteralOp : EmitC_Op<"literal", [Pure, CExpressionInterface]> {
```
> I think we need to overwrite hasSideEffects to return false because literals are always inlined on their own. See also my comment about a test for that.

Thanks. Then, I will allow myself to repeat the question from the above block, as it is more appropriate here. In the case where `literalOp` has a method `bool hasSideEffects() { return false; }` in `Emitc.td`, does the `LiteralOp` really need the `Pure` trait?

Vladislave0-0 commented:
> Literal is marked `pure`. Won't this expose the expression to CSE, causing the example above to be left without any operation with the expression? (see this [related discussion](https://github.com/llvm/llvm-project/pull/143485#issuecomment-2965378156))

Thanks! If I understood your question correctly, I think there should be no problem with CSE, because emitc.expression creates an isolated region that protects internal operations from CSE with external operations. So the test above will work correctly.

kchibisov commented:
What do you think about ignoring `literal` in `form-expression` pass? I'd say that the point of this PR is that if your `literal` somehow ended in `emitc.expression` due to other transformations, to not blow up, but generally, we don't need it in the expression itself, because it's inlined anyway, so we just make things a bit more messy.

Vladislave0-0 committed:
[mlir][emitc] Ignore literal in form-expression

Literal is always inlined, so there's no need to form expressions
with it as well.

jpienaar commented on mlir/lib/Dialect/EmitC/Transforms/FormExpressions.cpp:
```diff
@@ -38,7 +38,7 @@ struct FormExpressionsPass
       if (isa<emitc::CExpressionInterface>(*op) &&
           !op->getParentOfType<emitc::ExpressionOp>() &&
-          op->getNumResults() == 1)
+          !isa<emitc::LiteralOp>(op) && op->getNumResults() == 1)
```
I was wondering about the ordering here, I think op->getNumResults() may be cheaper than getParentOfType (as it could keep walking until top) and even checking for interface is linear scan compared to checking if isa LiteralOp. Probably not crazy significant time wise though.

jpienaar commented on mlir/lib/Target/Cpp/TranslateToCpp.cpp:
```diff
@@ -442,6 +443,16 @@ static LogicalResult printOperation(CppEmitter &emitter, emitc::LoadOp loadOp) {
 
+static LogicalResult printOperation(CppEmitter &emitter,
+                                    emitc::LiteralOp literalOp) {
+  // If literalOp is used inside an expression, we treat it as an embedded one.
```
inlined one?

Vladislave0-0 committed:
[mlir][emitc] Reorder match conditions for efficiency

Vladislave0-0 committed:
[mlir][emitc] Corrected the test

aniragil commented:
> @aniragil do I need to close this PR?

Let's first see if my PR is accepted (would be great if you could also take a look at it)
